docs/design/bin_pack_fault_domains.md

```diff
@@ -0,0 +1,126 @@
+# Support bin packing of fault domains
+
+## Metadata
+
+* Authors: @johscheuer
+* Created: 2021-10-01
+* Updated: 2021-10-01
+
+## Background
+
+In the current implementation of the operator we try to distribute our Pod across as many fault domains as possible.
+If a cluster has more fault domains than the operator creates Pods for a specific class all Pods will be on a different fault domain.
+This has the drawback that the operator can only interact on a single Pod for safe operations e.g. to minimize the risk of data loss.
+For clusters with a non-trivial amount of Pods this means that some operational tasks will take very long.
+A better way would be to allow the operator to bin pack Pods in the same fault domain.
+
+## General Design Goals
+
+* Provide a way how a user can specify the number of logical fault domains
+* The user should be able to decide if the number of logical fault domains is a requirement (default should be optional)
+
+## Current Implementation
+
+The current implementation adds a `preferredDuringSchedulingIgnoredDuringExecution` `PodAntiAffinity` to every Pod [code](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/master/internal/pod_models.go#L308-L334).
+The label selector will contain the process class and the cluster name, to ensure processes with the same class are distributed across different fault domains is possible.
+A user can define additional `PodAntiAffinity` but can't prevent the operator from createn the default `PodAntiAffinity`.
+The only expection is the specival fault domain `foundationdb.org/none` and `foundationdb.org/kubernetes-cluster`.
+
+## Proposed Design
+
+The idea would be to use a custom label like `foundationdb.org/distribution-key`.
+As value we would use the process number modulo the `DesiredFaultDomains` and a prefix for the `BinPack` mode.
```

Comment:
Could this produce scenarios where we had fewer logical fault domains than desired? For instance, if we had process groups with ids `storage-1`, `storage-2`, `storage-3`, with double replication and three fault domains desired, they would have the logical fault domains `1`, `2`, and `0`. If we then replaced `storage-2`, we would get the process groups `storage-1`, `storage-3`, `storage-4`, with logical fault domains `1`, `0`, and `1`.

Reply:
As an alternative, maybe we could look at how many current non-removed pods are in each logical fault domain, and assign each new pod to whichever fault domain has the fewest pods currently.

Reply:
That could also make the process of detecting changes to the desired number of fault domains easier, because we could look at fault domains that are overloaded compared to the desired distribution and trigger replacements for processes in those overloaded fault domains.

Reply:
Yes, that could happen and is stated below `The operator won't guarantee that the fault domains are equally used.` which in that case would lead to a non-functional cluster. I would rather put that as a limitation for that design. I think about the proposal with the assignment of the fault domains and add some more information in the design.